### PR TITLE
Add python dispatch trace offset constant

### PR DESCRIPTION
**Pull request by sypneiwski**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #84428
* #84427
* #83984
* #83983
* __->__ #83982

### CODE REVIEW

File: test/test_python_dispatch.py [hunk 1]
```diff
@@ -1821,5 +1823,28 @@ def __torch_dispatch__(cls, func, types, args, kwargs):
...
+        # If this test stops working because of a change in Python dispatch,
+        # please make sure to change the constant itself.
+        expected_offset = _PYTHON_DISPATCH_TRACE_OFFSET
+        tb = []
+        with MockDispatchMode(tb) as dispatch:
+            # Tests if the line calling the kernel launch is at the expected offset
+            # in the stack trace using the fact that the unique variable name will
+            # appear in it.
+            _long_and_unique_launch_name_ = torch.rand(3)
+            self.assertNotIn(
```
Comment (ngimel) [inline test/test_python_dispatch.py:1843]: but you are not checking that unique name is in the trace? (In the next PR you are doing it for cuda trace offset)

**Comment by ezyang** [comment]
A more robust way to do this is to just scan the trace until you get out of torch library code; it will save you from hardcoding a constant and is more robust.

File: test/test_python_dispatch.py (hunk 1 above)
Comment (sypneiwski) [reply test/test_python_dispatch.py:1843]: Sorry, fixed it now. The test did not work with CrossRef, I tried to work around it but in the end I decided skipping it is best.

**Comment by sypneiwski** [comment]
> A more robust way to do this is to just scan the trace until you get out of torch library code; it will save you from hardcoding a constant and is more robust.

I've been looking into that, but I feel there are cases where we might want to print some of the torch library code, for example if the user creates their own `__torch_dispatch__` or `__torch_function__` mode. I've also found it hard to do safely; solutions I've thought of such as searching for torch keywords in the filenames might not always work.

For these reasons I've initially decided to introduce the offsets. I'm also now wondering if it might make more sense to print the full stack trace, which will make the error messages less readable but will be safer.

**Comment by ezyang** [comment]
I'm cool with printing full trace

**Pull request closed without merging**